src/xdg.rs

```diff
@@ -332,7 +389,30 @@ impl IconLoader {
         // Resolve icon from name if it is not an absolute path.
         let mut path = Path::new(icon);
         if !path.is_absolute() {
-            path = self.icons.get(icon).ok_or(Error::NotFound)?;
+            // Get all available icons matching this icon name.
+            let matching_icons = self.icons.get(icon).ok_or(Error::NotFound)?;
+
+            // Sort icons by best match for the desired size.
+            let mut matching_icons: Vec<_> = matching_icons.iter().collect();
+            matching_icons.sort_by_key(|(&icon_type, _)| match icon_type {
+                ImageType::SizedBitmap(icon_size) => match icon_size.cmp(&size) {
+                    // If the size of the PNG is identical to ours, it’s the best option.
+                    Ordering::Equal => 0,
+                    // Otherwise order by closest size bigger than ours.
+                    Ordering::Greater => icon_size - size,
+                    // And only prefer smaller sizes after unknown sizes.
+                    Ordering::Less => 8192 + size - icon_size,
+                },
+                // Scalable is obviously the second best option to get crisp icons.
+                ImageType::Scalable => 1,
+                // Here we take the bet that an unknown file will be better than a known smaller.
+                ImageType::Bitmap => 8192,
+                // And finally, symbolic is the same icon but without colours, usually too dark.
+                ImageType::Symbolic => 16384,
+            });
```

Comment:
I applied all the changes that were pretty non-controversial to speed things up a bit, but with this specific code I'd like to get your feedback first.

First I'm curious what the purpose of sorting is. The "ideal" icon can always be determined in a single iteration, correct? Even to ensure the `vec` is sorted, you'd have to iterate over it anyway so it wouldn't help you. And since it's created from a `HashMap` it's rather unlikely that this is already sorted?

So if we just iterate over them all once, we can save ourselves the sorting and it should avoid having to re-allocate to go from `HashMap` to `Vec`.

Besides that, I also think that this would probably be a good separate method on `self`. It's so long due to the (useful) documentation, that having it in a separate method with a succinct name would likely make reading of this `load` function much easier.

Reply:
Thanks for the improvements, they are all good!

I have just pushed a new commit which goes a bit further into what I had in mind for the sort, basically doing it once at initial load time and then only sorting again when the size changes.  `IconLoader` isn’t aware of the size until `load()` so we currently have to keep doing the load every time, not only on size changes, but that’s for a future refactor.